sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -88,20 +88,42 @@ RT::PiProgram ProgramManager::getBuiltOpenCLProgram(OSModuleHandle M,
                                                     const context &Context) {
   std::shared_ptr<context_impl> Ctx = getSyclObjImpl(Context);
   std::map<OSModuleHandle, RT::PiProgram> &CachedPrograms =
-      Ctx->getCachedPrograms();
+      Ctx->acquireCachedPrograms();
+
   auto It = CachedPrograms.find(M);
   if (It != CachedPrograms.end())
     return It->second;
 
+  Ctx->releaseCachedPrograms();
+
+  // TODO improvement to employ after code-split feature is done:
+  // - put atomic pointer in program cache as data with nullptr before
+  //   the program is actually built
+  // - any thread which will find nullptr in cache will wait until the pointer
+  //   is not null anymore
+
   DeviceImage *Img = nullptr;
   using PiProgramT = remove_pointer_t<RT::PiProgram>;
   unique_ptr_class<PiProgramT, decltype(&::piProgramRelease)> ProgramManaged(
       loadProgram(M, Context, &Img),
       RT::PluginInformation.PiFunctionTable.piProgramRelease);
 
   build(ProgramManaged.get(), Img->BuildOptions);
-  RT::PiProgram Program = ProgramManaged.release();
-  CachedPrograms[M] = Program;
+
+  Ctx->acquireCachedPrograms();
+
+  using IterT = std::map<OSModuleHandle, RT::PiProgram>::iterator;
+
+  RT::PiProgram Program = ProgramManaged.get();
+  std::pair<IterT, bool> Inserted = CachedPrograms.insert(std::make_pair(M, Program));
```

Comment:
This is why `auto` has been invented. Welcome into the modern C++ world :-)

Reply:
What about just:
```
auto Inserted = CachedPrograms.insert({ M, Program });
```

Reply:
Done